[REFACTOR] Refresh Token 전달 방식 개선 및 보안 강화 about HttpOnly 및 Cookie

jnujh opened the pull request:
## 💡 작업 배경 및 문제점

기존 로그인 응답 방식은 **Access Token**과 **Refresh Token**을 모두 JSON 본문에 담아 클라이언트에게 전달했습니다. 하지만 프론트엔드 팀에서는 보안 강화를 위해 상대적으로 오래 보관되는 **Refresh Token**을 JavaScript에서 접근할 수 없는 **HttpOnly 쿠키**에 저장하고자 했습니다.

### 핵심 문제

**HttpOnly** 속성이 붙은 쿠키는 보안상의 이유로 브라우저가 JavaScript의 접근을 원천적으로 차단합니다. 따라서 프론트엔드에서는 쿠키에 저장된 Refresh Token을 JavaScript로 읽어서 Authorization 헤더에 담는 것이 기술적으로 불가능했습니다. 이는 현재 백엔드의 Access Token 재발급 API 명세(클라이언트가 Authorization 헤더에 Refresh Token을 담아 보내도록 되어 있음)와 충돌하는 문제였습니다.

### 보안적 중요성 (XSS 공격 방어)

이 문제의 핵심에는 **XSS(Cross-Site Scripting) 공격 방어**라는 중요한 보안 원칙이 있습니다. 만약 해커가 웹사이트에 악성 스크립트를 삽입하는 데 성공하면, JavaScript가 접근할 수 있는 모든 정보를 탈취할 수 있습니다. **HttpOnly 쿠키**는 브라우저에게 "이 쿠키는 절대 JavaScript가 만지지 못하게 해!"라고 명령하는 것과 같으므로, 만에 하나 XSS 공격을 당하더라도 해커가 가장 중요한 **Refresh Token**을 훔쳐가는 것을 막을 수 있습니다.

### 결론

현재 우리의 API 명세는 프론트엔드의 올바른 보안 전략과 충돌하고 있었으며, 백엔드 코드를 수정하여 프론트엔드가 **HttpOnly 쿠키**를 안전하게 사용할 수 있도록 지원해야 했습니다. 이 PR은 이러한 문제점을 해결하고 보안을 강화하기 위해 진행되었습니다.

---

## 🚀 작업 내용

프론트엔드의 보안 강화 요구사항에 따라, **Refresh Token**의 전달 방식을 기존 JSON 본문에서 **HttpOnly** 속성이 적용된 쿠키로 전면 변경했습니다. 이를 통해 **XSS(Cross-Site Scripting) 공격**으로부터 Refresh Token을 안전하게 보호합니다.

주요 변경 사항은 다음과 같습니다:

1. **로그인 & 회원가입 API 응답 변경**
    - 로그인/회원가입 성공 시, 서버는 Set-Cookie 헤더를 통해 **Refresh Token**을 **HttpOnly 쿠키**로 전달합니다.
    - 클라이언트 응답 본문(JSON)에서는 **Refresh Token** 필드를 제거하고, **Access Token**만 포함하도록 수정했습니다. 이를 위해 응답 전용 DTO(Body record)를 추가했습니다.
2. **토큰 재발급 API 요청/응답 방식 변경**
    - reissue API가 Authorization 헤더 대신, 브라우저가 자동으로 전송하는 쿠키에서 **Refresh Token**을 받도록 **@CookieValue** 를 사용해 수정했습니다.
    - 재발급 성공 시, 새로 발급된 **Refresh Token** (Refresh Token Rotation) 또한 **HttpOnly 쿠키**로 갱신하여 전달합니다.
3. **전역 예외 처리 강화**
    - 필수 쿠키(refreshToken)가 요청에 포함되지 않았을 때 발생하는 **MissingRequestCookieException** 을 처리하기 위한 핸들러를 **GlobalExceptionHandler** 에 추가했습니다. 이를 통해 500 에러 대신 명확한 **400 Bad Request**를 응답합니다.
4. **테스트 코드 수정 및 보강**
    - 변경된 서비스 로직에 맞춰 AuthServiceImplTest의 단위 테스트를 수정했습니다.
    - **@WebMvcTest** 를 사용하여 AuthController의 쿠키 처리 로직을 검증하는 통합 테스트(AuthControllerTest)를 신규 작성했습니다.
5. **설정 외부화 리팩토링**
    - AuthController에 하드코딩되어 있던 Refresh Token 만료 시간을 application.yml의 jwt.refresh-token-validity-in-seconds 값을 사용하도록 리팩토링했습니다.

---

## ⏱️ 소요 시간

3h

---

## 🤔 고민했던 내용

1. 응답 DTO를 어떻게 분리할 것인가?
    
    Refresh Token을 쿠키로 옮기면서, 클라이언트에게 보내는 JSON 응답에서는 해당 필드를 제거해야 했습니다. 처음에는 기존 LoginSuccessResponseDto에서 필드를 제거하는 방안을 생각했지만, 이 경우 AuthService가 생성한 Refresh Token을 AuthController로 전달할 방법이 사라지는 문제가 있었습니다.
    
    **해결:** 서비스와 컨트롤러 간의 데이터 전달 역할은 유지하되, 최종 응답 명세만 변경하기 위해 기존 DTO 내부에 **응답 본문 전용 Body record를 중첩하여 정의**하는 방식을 채택했습니다. 이를 통해 각 계층의 책임은 명확히 유지하면서 API 명세의 일관성을 확보할 수 있었습니다.
    
2. 프레임워크 레벨 예외는 어떻게 처리해야 하는가?
    
    @CookieValue를 사용하면서, 필수 쿠키가 누락되었을 때 MissingRequestCookieException 이 발생하며 테스트가 실패하는 문제를 마주했습니다. 저희 팀의 예외 처리 컨벤션은 커스텀 예외를 던져서 처리하는 것이었지만, 이 경우에는 적용할 수 없었습니다.
    
    **이유:** 해당 예외는 저희가 작성한 컨트롤러 메소드의 코드가 실행되기 전, **Spring의 ArgumentResolver 단에서 발생하는 프레임워크 레벨 예외**이기 때문입니다. 즉, 저희 코드 내에서 try-catch로 잡아 커스텀 예외로 변환할 기회 자체가 없었습니다.
    
    **해결:** 이러한 프레임워크 레벨의 예외를 처리하는 표준적인 방법인 **@RestControllerAdvice(GlobalExceptionHandler)에 직접 핸들러를 추가**하는 방식을 채택했습니다. 이를 통해 클라이언트의 잘못된 요청에 대해 명확한 **400 Bad Request** 응답을 제공하고, 애플리케이션의 안정성을 높였습니다.
    
3. 무엇을, 어떻게 테스트해야 하는가?
    
    단순히 서비스 로직의 단위 테스트만 수정하는 것은 충분하지 않다고 판단했습니다. 이번 작업의 핵심은 HTTP 프로토콜 수준에서의 변화(쿠키, 헤더, 응답 본문)였기 때문입니다.
    
    **해결:** **MockMvc**를 사용한 **@WebMvcTest** 를 적극적으로 활용하여 AuthController에 대한 통합 테스트를 작성했습니다. 이를 통해 실제 컨트롤러가 HTTP 요청을 받았을 때, 의도한 대로 쿠키의 속성(HttpOnly, Secure등)이 올바르게 설정되는지, 응답 본문에서 민감한 정보가 잘 제거되었는지 등을 꼼꼼하게 검증할 수 있었습니다.
    

---

## 💬 리뷰 중점사항

- 변경된 인증/인가 흐름(로그인, 재발급)이 보안적인 관점에서 올바르게 구현되었는지 확인 부탁드립니다.
- AuthController에서 ResponseCookie를 생성하고 사용하는 방식이 적절한지 검토해 주세요.
- GlobalExceptionHandler에 추가된 예외 처리 로직이 다른 핸들러와 충돌 없이 잘 동작할지 확인 부탁드립니다.
- 새롭게 작성된 AuthControllerTest가 쿠키 기반 인증의 핵심적인 시나리오들을 충분히 커버하고 있는지 검토해 주세요.

---

## 🔗 관련 이슈

- Close #153

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## 릴리스 노트

* **개선 사항**
  * 새로고침 토큰이 HttpOnly/Secure 쿠키로 발급·갱신됩니다.
  * 로그인·회원가입·토큰재발급 응답에 상태와 accessToken을 담은 표준화된 응답 본문이 포함됩니다 (응답 본문에서 refreshToken 노출 없음).
  * 토큰 재발급 요청이 쿠키 기반으로 처리됩니다.
  * 누락된 쿠키에 대한 400 에러 처리(명확한 오류 메시지) 추가.

* **테스트**
  * 인증 흐름(로그인·회원가입·재발급)과 쿠키 동작을 검증하는 통합/단위 테스트 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jnujh committed:
feat: 로그인 응답 본문 전용 DTO 추가 (#153)

Refresh Token을 HttpOnly 쿠키로 전달하는 방식으로 변경함에 따라,
클라이언트에게 보내는 JSON 응답 본문에서는 Refresh Token을 제외합니다.
이를 위해 기존 LoginSuccessResponseDto 내부에 accessToken만 포함하는 중첩 record 'Body'를 새로 정의했습니다.

jnujh committed:
feat: 로그인 시 Refresh Token을 HttpOnly 쿠키로 전달하도록 변경 (#153)

로그인 및 회원가입 성공 시, Refresh Token의 탈취 위험(XSS)을 줄이기 위해 전달 방식을 JSON 본문에서 HttpOnly 쿠키로 변경합니다.

주요 변경 사항:
- kakaoLogin, register 메소드 수정
- Spring의 ResponseCookie 빌더를 사용하여 httpOnly, secure 속성이 적용된 Refresh Token 쿠키 생성
- 생성된 쿠키를 'Set-Cookie' 헤더에 담아 응답
- 응답 본문에는 accessToken만 포함된 LoginSuccessResponseDto.Body를 사용
이 변경으로 프론트엔드는 JavaScript로 접근 불가능한 안전한 공간에 Refresh Token을 저장할 수 있게 됩니다.

jnujh committed:
feat: 토큰 재발급 응답 본문 전용 DTO 추가 (#153)

Refresh Token 재발급 시, 새로 발급된 Refresh Token을 HttpOnly 쿠키로 전달하는 방식으로 변경함에 따라,
클라이언트에게 보내는 JSON 응답 본문에서는 Refresh Token을 제외합니다.
이를 위해 기존 ReissueResponseDto 내부에 accessToken만 포함하는 중첩 record 'Body'를 새로 정의했습니다.

jnujh committed:
refactor: reissue 메소드 Refresh Token 처리 방식 변경 (#153)

AuthController에서 Refresh Token을 HttpOnly 쿠키로부터 직접 받아 전달하도록 변경됨에 따라,
AuthService의 reissue 메소드 시그니처 및 내부 로직을 수정합니다.
주요 변경 사항:
- reissue 메소드의 파라미터를 'bearerToken'에서 순수한 'refreshToken'으로 변경
- 'jwtProvider.extractToken()' 호출 제거 (더 이상 "Bearer " 접두사를 처리할 필요 없음)이 변경으로
컨트롤러와 서비스 간의 Refresh Token 전달 방식이 명확해지고, 불필요한 토큰 추출 로직이 제거됩니다.

jnujh committed:
feat: Access Token 재발급 시 Refresh Token을 HttpOnly 쿠키로 전달 (#153)

Access Token 재발급 API(/api/auth/reissue) 호출 시, Refresh Token을 Authorization 헤더 대신
HttpOnly 쿠키를 통해 받도록 변경하고, 새로 발급된 Refresh Token 또한 HttpOnly 쿠키로 응답합니다.
주요 변경 사항:
- reissue 메소드의 @RequestHeader("Authorization")을 @CookieValue("refreshToken")으로 변경
- AuthService로부터 받은 새로운 Refresh Token을 ResponseCookie에 담아 'Set-Cookie' 헤더로 응답
- 응답 본문에는 새로운 Access Token만 포함된 ReissueResponseDto.Body를 사용
- Swagger API 문서의 응답 스키마도 ReissueResponseDto.Body로 업데이트
이 변경으로 Refresh Token의 보안이 강화되며, XSS 공격으로부터 안전하게 보호됩니다.

jnujh committed:
test: AuthService의 reissue 테스트 코드 수정 (#153)

AuthService의 reissue 메소드 시그니처가 'bearerToken'을 받는 방식에서 순수한 'refreshToken'을 받도록 변경됨에 따라, 관련 단위 테스트 코드를 수정합니다.
주요 변경 사항:
- "Bearer " 접두사를 사용하던 로직을 제거하고, 순수한 토큰 문자열을 사용하도록 변경
- 불필요해진 jwtProvider.extractToken() Mocking 로직 삭제
이제 테스트는 변경된 메소드 시그니처에 맞춰 순수한 refreshToken을 직접 전달하여 reissue 로직을 검증합니다

jnujh committed:
feat: 필수 쿠키 누락(MissingRequestCookieException) 예외 처리 핸들러 추가 (#153)

@CookieValue 어노테이션을 사용하는 컨트롤러에서 필수 쿠키가 누락되었을 때 발생하는 MissingRequestCookieException을
전역적으로 처리하기 위한 핸들러를 추가합니다.
이전에는 이 예외가 처리되지 않아 500 Internal Server Error를 유발할 수 있었습니다.
새로운 핸들러는 해당 예외 발생 시, ErrorCode.INVALID_INPUT_VALUE를 사용하여 HTTP 400 Bad Request와 일관된 형식의 에러 메시지를 반환합니다.
이를 통해 클라이언트의 잘못된 요청에 대해 명확한 피드백을 제공하고, 관련 테스트 코드가 의도대로 동작하도록 보장합니다.

jnujh committed:
test: AuthController 쿠키 처리 로직 테스트 작성 (#153)

Refresh Token을 HttpOnly 쿠키로 처리하는 로직을 검증하기 위해 AuthController에 대한 테스트를 추가합니다.
@WebMvcTest와 MockMvc를 사용하여 실제와 유사한 HTTP 요청/응답을 시뮬레이션하고, 컨트롤러의 쿠키 생성 및 수신 기능을 검증합니다.
주요 검증 항목:
- 로그인/회원가입 시, 응답 헤더에 HttpOnly, Secure 속성을 가진 refreshToken 쿠키가 포함되는지 확인
- 재발급 요청 시, 요청 쿠키를 @CookieValue가 올바르게 수신하는지 확인
- 재발급 응답 시, 새로운 refreshToken 쿠키로 갱신되는지(Rotation) 확인
- 모든 응답 본문에서 refreshToken 필드가 제거되었는지 확인
- 필수 쿠키가 없을 때 400 Bad Request가 발생하는지 확인

jnujh committed:
refactor: Refresh Token 만료 시간 설정 외부화 (#153)

ji-mim commented on src/test/java/com/kakaotech/team18/backend_server/domain/auth/controller/AuthControllerTest.java:
```diff
@@ -0,0 +1,169 @@
+        // when & then
+        mockMvc.perform(post("/api/auth/reissue")) // 쿠키 없이 요청
+                .andExpect(status().isBadRequest()); // @CookieValue는 쿠키가 없으면 400 Bad Request 반환
+    }
```
HTTP 쿠키에 RefreshToken이 아니라 다른 Token이 들어있는 경우는 없겠죠?

jnujh replied:
다른 Token이 들어가는건 고려하지 않고 구현했습니다.
이후에 별도의 토큰을 HTTP 쿠키에서 관리하게 된다면 적절한 조치가 필요할 거 같습니다 !

ji-mim commented on src/main/java/com/kakaotech/team18/backend_server/domain/auth/dto/LoginSuccessResponseDto.java:
```diff
@@ -18,5 +18,14 @@ public record LoginSuccessResponseDto(
-    // LoginResponse 인터페이스를 구현합니다.
+        @Schema(description = "로그인 성공 시 실제 클라이언트에게 전달되는 응답 본문")
+        public record Body(
+                @Schema(description = "응답 상태", example = "LOGIN_SUCCESS")
```
LoginSuccessResponseDto만 return 하는 경우는 없고 Body record만 return 되는 것 같은데 혹시 적어두신  

```
AuthService가 생성한 Refresh Token을 AuthController로 전달할 방법이 사라지는 문제가 있었습니다.
```

와 관련해서 조금만 더 설명해주실 수 있으실까요? 
그리고 클라이언트에게 `clubIdAndRoleList` 값도 전달해주지 않고 있는 것 같아 보이는데 아닐까요?

jnujh replied:
1. LoginSuccessResponseDto 구조와 Refresh Token 전달 문제에 대한 답변:
먼저 LoginSuccessResponseDto가 직접 반환되지 않고 그 안의 Body만 반환되는 것처럼 보이는 이유는, 
Service와 Controller의 역할을 명확히 분리하기 위해서입니다.
•AuthService의 역할: 
인증에 필요한 모든 데이터를 생성하고 모아서 전달하는 것입니다. 
여기에는 클라이언트에게 전달될 accessToken, clubIdAndRoleList 뿐만 아니라, 
Controller가 HTTP 응답(Cookie)에 담아줘야 할 refreshToken도 포함됩니다.

•AuthController의 역할: 
AuthService로부터 데이터를 받아서, 어떤 정보는 HTTP Body에 담고, 
어떤 정보는 HTTP Cookie나 Header에 담을지 결정하여 최종 응답을 만드는 것입니다.
만약 AuthService가 클라이언트에게 보낼 정보(accessToken, clubIdAndRoleList)만 반환한다면, 
AuthController는 refreshToken을 쿠키에 담을 방법이 사라집니다.
그래서 LoginSuccessResponseDto라는 'Service와 Controller 사이의 데이터 운반용 객체'를 만들었습니다. 
이 객체는 다음과 같은 정보를 모두 가지고 있습니다.

•accessToken (Body로 전달)
•clubIdAndRoleList (Body로 전달)
•refreshToken (Cookie로 전달)

AuthController는 AuthService로부터 LoginSuccessResponseDto를 통째로 받은 뒤, 
refreshToken은 꺼내서 HttpOnly 쿠키에 설정하고, 나머지 정보(accessToken, clubIdAndRoleList)는 Body에 담아 클라이언트에게 응답하는 구조입니다.
따라서 지민님께서 보신 것처럼 클라이언트 입장에서는 LoginSuccessResponseDto의 Body 부분만 받는 것이 맞습니다. LoginSuccessResponseDto 자체는 서버 내부(Service -> Controller)에서만 사용되는 DTO입니다.

2. clubIdAndRoleList 전달 누락에 대하여이 부분
제가 실수한 것 같습니다! 
지적해주신 대로 LoginSuccessResponseDto를 생성할 때 clubIdAndRoleList를 포함시키지 않았네요. 
클라이언트가 로그인 직후 어떤 동아리에 속해있는지, 그리고 권한이 무엇인지 알아야 원활한 서비스 이용이 가능하기 때문에 이 값은 꼭 전달되어야 합니다.
이 문제를 해결하기 위해 LoginSuccessResponseDto 생성자에 clubIdAndRoleList를 추가하고, kakaoLogin과 register 메서드에서 이 값을 채워서 반환하도록 코드를 수정하도록 하겠습니다

ji-mim reviewed (approved):
지훈님 수고 많으셨습니다!
클라이언트 쪽에 json 응답 값으로 refresh token을 바로 전달해 외부에 노출되지 않도록 http cookie를 사용해 이를 전달하고 reissue 시에도 쿠키에서 refresh token 값을 추출해서 사용하는 흐름인 것 같은데 맞을까요?  

코멘트 한번씩만 확인해주세요 수고하셨습니다.

jnujh commented:
클라이언트 쪽에 json 응답 값으로 refresh token을 바로 전달해 외부에 노출되지 않도록 http cookie를 사용해 이를 전달하고 reissue 시에도 쿠키에서 refresh token 값을 추출해서 사용하는 흐름인 것 같은데 맞을까요?
->
넵 맞습니다 !

jnujh committed:
feat: 로그인/회원가입 응답에 clubIdAndRoleList 추가 (#153)

로그인 및 회원가입 성공 시 클라이언트에게 전달되는 JSON 응답에
사용자의 동아리 목록 및 역할 정보(`clubIdAndRoleList`)를 포함하도록 수정합니다.

기존에는 `AuthService`가 해당 정보를 생성하여 `AuthController`까지는 전달했으나,
최종적으로 클라이언트에게 보내는 응답 DTO(`LoginSuccessResponseDto.Body`)에 필드가 누락되어 정보가 전달되지 않는 문제가 있었습니다.

이 변경으로 클라이언트는 로그인 후 별도의 API 호출 없이 사용자의 동아리 정보를 획득할 수 있습니다.

gary5876 reviewed (approved):
수정사항 확인했습니다. 고생하셨어요 👍
